dFusion/dFusionSpec.md

```diff
@@ -6,34 +6,72 @@ A specification developed by Gnosis.
+All changes to the state will be announced via smart contract events.
+Thus, the full state will be fully reproducible for any participant by replaying all blocks since the creation of the smart contract.
+The following diagram shows how the "compressed" state hash is constructed:
 
 ![State construction](./dFusion%20rootHash.png?raw=true "State construction")
 
-To allow `K` to be small, a bi-map of an accounts public key (on-chain address) to its `accountIndex` will be stored in the anchor contract as well. Accounts will pay "rent" to occupy an active account. The account index can be used to locate a users token balances in the state.
+To allow `K` to be small, a bi-map of an accounts public key (on-chain address) to its `accountIndex` will be stored in the anchor contract as well. 
+Accounts will pay "rent" to occupy an active account. The account index can be used to locate a users token balances in the state.
 
-All orders are encoded as limit sell orders: `(accountIndex, fromTokenIndex, toTokenIndex, limitPrice, amount, batchId, signature)`.
-The order should be read in the following way: the user occupying the specified *accountIndex* would like to sell the token *fromTokenIndex* for *toTokenIndex* for at most the *limitPrice* and the *amount* specified.
-The *batchId* and *signature* allow a third party to submit an order on behalf of others (saving gas when batching multiple orders together).
-The user only has to specify which batch their order is valid for and sign all the information with their private key.
+Furthermore, we store a bi-map of token address to its index `0 <= t <= T`, for each token that can be traded on the exchange.
+When specifying tokens inside orders, deposits and withdrawel requests, we use the token's index `0 <= t <= T` instead of the full address.
+
+As orders, deposits and withdrawl requests are collected they are not directly stored in the smart contract.
+Doing so would require a `SSTORE` instruction which would be too gas-expensive.
+Assuming an order can be encoded in 256 bits, storing a batch of 10.000 on chain would cost ~5M gas (5.000 gas for SSTORE * 10.000 orders).
+Instead the smart contract emits a smart contract event containing the relevant order information (account, from_token, to_token, limit) and stores a rolling SHA hash.
+For a new order, the rolling hash is computed by hashing the previous rolling hash with the current order.
+
+// TODO: Pending withdrawels
+
+## Batch Auction Workflow
 
 The trading workflow consists of the following sequential processes:
+0. Account opening, deposits & withdrawels
 1. On-Chain order collection
 2. Transition function from sha to Pedersen hashes (zkSnark)
 3. Finding the batch price: optimization of batch trading surplus (offchain)
 4. Verifying batch price and trade execution (zkSnark)
 5. Processing of pending exits and deposits (zkSnark)
-6. Restart with step 1
+6. Restart with step 0
+
+### Account opening, deposits & withdrawels
+
+// TODO
```

Comment:
I suppose it wouldn't hurt for me to fill in as much of this in as possible before this is merged?

Reply:
I would keep the deposit and exit descriptions separate from the trade description.
It does not necessarily belong together.